app/lib/common/widgets/html_editor/components/mention_list.dart

```diff
@@ -0,0 +1,281 @@
+import 'package:acter/common/providers/room_providers.dart';
+import 'package:acter/common/widgets/html_editor/components/mention_item.dart';
+import 'package:acter_avatar/acter_avatar.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+import 'package:acter/common/widgets/html_editor/models/mention_attributes.dart';
+import 'package:acter/common/widgets/html_editor/models/mention_type.dart';
+import 'package:acter/features/chat_ng/providers/chat_room_messages_provider.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class MentionList extends ConsumerStatefulWidget {
+  const MentionList({
+    super.key,
+    required this.editorState,
+    required this.roomId,
+    required this.mentionType,
```

Comment:
this widget still feels a bit odd to me. In general, when we have widgets that are generalized but are then supposed to be doing slightly different things depending on _one_ flag passed in (like in this case, pass the mentions type to a different provider, that then just switches again on the same type), it seems like there is a miss for generalization and specialization here.

This isn't necessarily something we have to do before the merge, but it should be done either way.

Let's imagine a `GenericMentionList` of:
```dart
class GenericMentionList extends ConsumerStatefulWidget {

   final EditorState editorState;
   final VoidCallback onDismiss;

   // this is the incoming provider
   final ProviderBase<Map<String, String>> mentionsProvider;

   // these are the missing widgets
   final String headerTitle;
   final String notFoundTitle;

   const GenericMentionList({
      super.key,
      required this.editorState,
      required this.onDismiss
      required this.mentionsProvider,
      required this.headerTitle,
      required this.notFoundTitle,
});
 // ...
}
```

And then we'd have two Widget, each with its specifics:

```dart
class UserMentionList extends ConsumerWidget {
   final EditorState editorState;
   final VoidCallback onDismiss;
   final String roomId;

   const GenericMentionList({
      super.key,
      required this.editorState,
      required this.onDismiss,
      required this.roomId,
});
   @override
   Widget build(BuildContext context, WidgetRef ref) => GenericMentionList(
      editorState: editorState,
      onDismiss: onDismiss,
      // the actual provider
      mentionsProvider: userMentiothat is decided between, where that `MentionType` first comes into nProvider(roomId),
      // the fields
      headerTitle: L10n.of(context).users,
      notFoundTitle: l10n.of(context).noUsersFoundTitle,
    );
}


class RoomMentionList extends ConsumerWidget {
   final EditorState editorState;
   final VoidCallback onDismiss;
   final String roomId;

   const GenericMentionList({
      super.key,
      required this.editorState,
      required this.onDismiss,
      required this.roomId,
});
   @override
   Widget build(BuildContext context, WidgetRef ref) => GenericMentionList(
      editorState: editorState,
      onDismiss: onDismiss,
      // the actual provider
      mentionsProvider: roomMentionProvider(roomId),
      // the fields
      headerTitle: L10n.of(context).chats,
      notFoundTitle: l10n.of(context).noChatsFoundTitle,
   );
}
````


And then, whatever first encounters that `MentionType` would just pick between these two higher level widgets and render that. Where each one has its specific neat and clean stuff together and the inner one doesn't have these nasty switch statement everyone. Also, we get rid of one more provider that has a switch, too...

Reply:
after writing it I realised, these don't even need to be consumers, just stateless widgets is enough because we need the L10n.of(Context) there...

Reply:
Yeah I can see how this change will improve code cleaning and too many widgets wrapping. Nice thinking!